File: Data Structures & Algorithms/combination-target-sum/submission-2.py
from typing import List

class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        ans = []
        subset = []
        def dfs(i: int):
            if sum(subset) == target:
                ans.append(subset.copy())
                return
            elif sum(subset) > target:
                return
            
            if i >= len(nums):
                return
            
            subset.append(nums[i])
            dfs(i)
            
            subset.pop()
            dfs(i + 1)
        dfs(0)
        return ans

File: Data Structures & Algorithms/three-integer-sum/submission-2.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> triplets;
        int n = nums.size();
        sort(nums.begin(), nums.end());
        for (int i = 0; i < n - 2; i++) {
            if (nums[i] > 0) {
                break;
            }
            if (i > 0 && nums[i] == nums[i-1]) {
                continue;
            }
            int first = i + 1;
            int end = n - 1;
            while (first < end) {
                int sum = nums[first] + nums[end];
                if (sum == -1 * nums[i]) {
                    triplets.push_back({nums[i], nums[first], nums[end]});
                    end--;
                    first++;
                    while (first < end) {
                        if (nums[end] == nums[end + 1]) {
                            end--;
                        }
                        else {
                            break;
                        }
                    }
                }
                else if (sum > -1 * nums[i]) {
                    end--;
                }
                else {
                    first++;
                }
            }
        }   
        return triplets;     
    }
};
